File: src/main/java/dev/salmon/keystrokes/hud/GuiKey.java
package dev.salmon.keystrokes.hud;

import dev.salmon.keystrokes.config.KeystrokesConfig;
import dev.salmon.keystrokes.hud.api.KeyRenderer;
import dev.salmon.keystrokes.hud.api.KeyStyle;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.settings.GameSettings;
import net.minecraft.client.settings.KeyBinding;
import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;

public class GuiKey extends Gui {
    protected final KeyStyle style = KeyStyle.INSTANCE;

    protected final FontRenderer fr = Minecraft.getMinecraft().fontRendererObj;
    protected final KeyBinding keyBinding;

    protected float relX, relY, width, height;
    protected boolean isPressed;
    protected float percentFaded;
    private long lastPress;

    public GuiKey(float x, float y, float width, float height, KeyBinding keyBinding) {
        this.relX = x;
        this.relY = y;
        this.width = width;
        this.height = height;
        this.keyBinding = keyBinding;
        this.percentFaded = 1.0F;
        this.lastPress = System.currentTimeMillis();
    }

    public void setLayout(float height, float relY) {
        this.height = height;
        this.relY = relY;
    }

    public void updateState() {
        boolean nowDown = isKeyDown(keyBinding.getKeyCode());
        if (nowDown != isPressed) {
            isPressed = nowDown;
            lastPress = System.currentTimeMillis();
            percentFaded = 0.0F;
        }
        long dt = System.currentTimeMillis() - lastPress;
        percentFaded = Math.min(1.0F, dt / (float) style.getFadingTime());
    }

    public void render(float baseX, float baseY, float scale) {
        float x = baseX + relX * scale;
        float y = baseY + relY * scale;
        float w = width * scale;
        float h = height * scale;

        KeyRenderer.drawBackground(x, y, w, h, style, isPressed, percentFaded, scale);
        KeyRenderer.drawCenteredText(fr, getKeyName(), x, y, w, h, style, isPressed, percentFaded);
    }

    protected boolean isKeyDown(int code) {
        return (code < 0) ? Mouse.isButtonDown(code + 100) : Keyboard.isKeyDown(code);
    }

    protected String getKeyName() {
        if (KeystrokesConfig.keystrokesElement.arrows) {
            GameSettings gs = Minecraft.getMinecraft().gameSettings;
            if (keyBinding == gs.keyBindForward) return "▲";
            if (keyBinding == gs.keyBindBack)    return "▼";
            if (keyBinding == gs.keyBindLeft)    return "◀";
            if (keyBinding == gs.keyBindRight)   return "▶";
        }
        int code = this.keyBinding.getKeyCode();
        switch (code) {
            case -100: return "LMB";
            case -99:  return "RMB";
            case -98:  return "MMB";
            case 200:  return "U";
            case 203:  return "L";
            case 205:  return "R";
            case 208:  return "D";
            case 210:  return "INS";
            case 29:   return "LCTRL";
            case 157:  return "RCTRL";
            case 56:   return "LALT";
            case 184:  return "RALT";
        }
        if (code >= 0 && code <= 223) return Keyboard.getKeyName(code);
        if (code >= -100 && code <= -84) return Mouse.getButtonName(code + 100);
        return "[]";
    }
}

File: src/main/kotlin/org/polyfrost/keystrokes/hud/GuiKey.kt
package org.polyfrost.keystrokes.hud

import net.minecraft.client.Minecraft
import net.minecraft.client.gui.FontRenderer
import net.minecraft.client.gui.Gui
import net.minecraft.client.settings.KeyBinding
import org.lwjgl.input.Keyboard
import org.lwjgl.input.Mouse
import org.polyfrost.keystrokes.config.KeystrokesConfig
import org.polyfrost.keystrokes.hud.api.KeyRenderer
import org.polyfrost.keystrokes.hud.api.KeyStyle
import kotlin.math.min

open class GuiKey(
    protected var relX: Float,
    var relY: Float,
    protected var width: Float,
    protected var height: Float,
    protected val keyBinding: KeyBinding
) : Gui() {
    protected val style: KeyStyle = KeyStyle

    protected val fr: FontRenderer = Minecraft.getMinecraft().fontRendererObj

    protected var isPressed: Boolean = false

    protected var percentFaded: Float = 1.0f
    private var lastPress: Long

    init {
        this.lastPress = System.currentTimeMillis()
    }

    fun setLayout(height: Float, relY: Float) {
        this.height = height
        this.relY = relY
    }

    fun updateState() {
        val nowDown = isKeyDown(keyBinding.keyCode)
        if (nowDown != isPressed) {
            isPressed = nowDown
            lastPress = System.currentTimeMillis()
            percentFaded = 0.0f
        }
        val dt = System.currentTimeMillis() - lastPress
        percentFaded = min(1.0f, dt / style.fadingTime.toFloat())
    }

    open fun render(baseX: Float, baseY: Float, scale: Float) {
        val x = baseX + relX * scale
        val y = baseY + relY * scale
        val w = width * scale
        val h = height * scale

        KeyRenderer.drawBackground(x, y, w, h, style, isPressed, percentFaded, scale)
        KeyRenderer.drawCenteredText(fr, this.keyName, x, y, w, h, style, isPressed, percentFaded)
    }

    protected open fun isKeyDown(code: Int): Boolean {
        return if (code < 0) Mouse.isButtonDown(code + 100) else Keyboard.isKeyDown(code)
    }

    private val keysMap = mapOf(
        -100 to "LMB",
        -99 to "RMB",
        -98 to "MMB",
        200 to "U",
        203 to "L",
        205 to "R",
        208 to "D",
        210 to "INS",
        29 to "LCTRL",
        157 to "RCTRL",
        56 to "LALT",
        184 to "RALT"
    )

    protected open val keyName: String?
        get() {
            val gs = Minecraft.getMinecraft().gameSettings

            if (KeystrokesConfig.keystrokesElement.arrows) {
                return when (keyBinding) {
                    gs.keyBindForward -> "▲"
                    gs.keyBindBack -> "▼"
                    gs.keyBindLeft -> "◀"
                    gs.keyBindRight -> "▶"
                    else -> null
                }
            }

            val code = keyBinding.keyCode

            keysMap[code]?.let { return it }

            return when (code) {
                in 0..223 -> Keyboard.getKeyName(code)
                in -100..-84 -> Mouse.getButtonName(code + 100)
                else -> "[]"
            }
        }
}